api/extension-guides/language-model-tutorial.md

```diff
@@ -9,13 +9,324 @@ MetaDescription: Tutorial that walks you through creating a VS Code extension th
+
+<!-- TODO: Add "Group" screenshot -->
+![]()
+
+At this point, you can run the extension from the "Run and Debug" view in the sidebar. This will open a new VS Code instance with the extension installed. If you open a project that contains code files and then open a code file, you should see the "Tutor" button in the title bar of the editor. You'll also see a "Tutor" option in the Command Palette.
+
+However, if you click on the "Tutor" button or select the option from the Command Palette, you'll see an error message that says "Command 'code-tutor.annotate' resulted in an error (command 'code-tutor.annotate' not found)". That's because we haven't implemented the command yet, so let's do that next.
+
+## Step 3: Implement the "Tutor" command
+
+Commands are defined in the `package.json` file, but they are implemented in the `src/extension.ts` file. Open this file and locate the following line.
+
+```ts
+const disposable = vscode.commands.registerCommand('code-tutor.helloWorld', () => {
+```
+
+Remember that we changed this command in the `package.json` file to be `code-tutor.annotate` instead of `code-tutor.helloWorld`. We also want to register a TextEditor command instead of just a command. This will give us a reference to the current text editor.
+
+Modify the line so that it looks like this:
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+```
+
+You can now run the extension again, and this time when you click on the "Tutor" button or select the option from the Command Palette, you should see a message that says "Hello from Code Tutor!".
+
+We're now ready to start building the functionality to generate AI-powered code annotations.
+
+## Step 4
+
+To get our Code Tutor annotations working, we'll need to do a few things...
+
+1. Get the code with line numbers from the current editor.
+2. Send the code to the Language Model API to get the annotations.
+3. Parse the annotations and display them in the editor.
+
+Let's start by getting the code from the current editor. Add the following method directly above the `export function deactivate() { }` line at the bottom of the `extension.ts` file.
+
+```ts
+function getCodeWithLineNumbers(textEditor: vscode.TextEditor) {
+ let currentLine = textEditor.visibleRanges[0].start.line;
+ const endLine = textEditor.visibleRanges[0].end.line;
+ let code = '';
+ while (currentLine < endLine) {
+  code += `${currentLine + 1}: ${textEditor.document.lineAt(currentLine).text} \n`;
+  currentLine++;
+ }
+ return code;
+}
+```
+
+This code uses the `visibleRanges` property of the TextEditor to get the lines that are currently visible in the editor. It then loops through these lines and adds each one to a string, along with the line number. Finally, it returns the string that contains the code with line numbers.
```

Comment:
Explain why we need the line number here! To use it to easily map the response of the Language Model.